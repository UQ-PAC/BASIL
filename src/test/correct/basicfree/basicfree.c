#include <stdlib.h>

void main() {
    int *foo = malloc(1 * sizeof(int));
    // foo -> alloc-1

    *foo = 1;
    // foo -> alloc-1

    free(foo);
    // foo -> nothing?
}