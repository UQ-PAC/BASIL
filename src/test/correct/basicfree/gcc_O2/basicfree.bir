000005d2: program
000005a0: sub __cxa_finalize(__cxa_finalize_result)
000005d3: __cxa_finalize_result :: out u32 = low:32[R0]

00000344:
0000043c: R16 := 0x10000
00000443: R17 := mem[R16 + 0xFB8, el]:u64
00000449: R16 := R16 + 0xFB8
0000044e: call R17 with noreturn

000005a1: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
000005d4: __do_global_dtors_aux_result :: out u32 = low:32[R0]

000002b1:
000002b5: #3 := R31 - 0x20
000002bb: mem := mem with [#3, el]:u64 <- R29
000002c1: mem := mem with [#3 + 8, el]:u64 <- R30
000002c5: R31 := #3
000002cb: R29 := R31
000002d3: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002d8: R19 := 0x11000
000002df: R0 := pad:64[mem[R19 + 0x10]]
000002e6: when 31:0[R0] <> 0 goto %000002e4
000005c8: goto %0000030d

0000030d:
00000310: R0 := 0x10000
00000317: R0 := mem[R0 + 0xFE0, el]:u64
0000031d: when R0 = 0 goto %0000031b
000005c9: goto %00000334

00000334:
00000337: R0 := 0x11000
0000033e: R0 := mem[R0 + 8, el]:u64
00000343: R30 := 0x730
00000346: call @__cxa_finalize with return %0000031b

0000031b:
00000323: R30 := 0x734
00000325: call @deregister_tm_clones with return %00000327

00000327:
0000032a: R0 := 1
00000332: mem := mem with [R19 + 0x10] <- 7:0[R0]
000005ca: goto %000002e4

000002e4:
000002ee: R19 := mem[R31 + 0x10, el]:u64
000002f5: R29 := mem[R31, el]:u64
000002fa: R30 := mem[R31 + 8, el]:u64
000002fe: R31 := R31 + 0x20
00000303: call R30 with noreturn

000005a5: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
000005d5: __libc_start_main_main :: in u64 = R0
000005d6: __libc_start_main_arg2 :: in u32 = low:32[R1]
000005d7: __libc_start_main_arg3 :: in out u64 = R2
000005d8: __libc_start_main_auxv :: in out u64 = R3
000005d9: __libc_start_main_result :: out u32 = low:32[R0]

0000020a:
00000426: R16 := 0x10000
0000042d: R17 := mem[R16 + 0xFB0, el]:u64
00000433: R16 := R16 + 0xFB0
00000438: call R17 with noreturn

000005a6: sub _fini(_fini_result)
000005da: _fini_result :: out u32 = low:32[R0]

0000001f:
00000025: #0 := R31 - 0x10
0000002b: mem := mem with [#0, el]:u64 <- R29
00000031: mem := mem with [#0 + 8, el]:u64 <- R30
00000035: R31 := #0
0000003b: R29 := R31
00000042: R29 := mem[R31, el]:u64
00000047: R30 := mem[R31 + 8, el]:u64
0000004b: R31 := R31 + 0x10
00000050: call R30 with noreturn

000005a7: sub _init(_init_result)
000005db: _init_result :: out u32 = low:32[R0]

000004f2:
000004f8: #5 := R31 - 0x10
000004fe: mem := mem with [#5, el]:u64 <- R29
00000504: mem := mem with [#5 + 8, el]:u64 <- R30
00000508: R31 := #5
0000050e: R29 := R31
00000513: R30 := 0x590
00000515: call @call_weak_fn with return %00000517

00000517:
0000051c: R29 := mem[R31, el]:u64
00000521: R30 := mem[R31 + 8, el]:u64
00000525: R31 := R31 + 0x10
0000052a: call R30 with noreturn

000005a8: sub _start(_start_result)
000005dc: _start_result :: out u32 = low:32[R0]

000001cb:
000001d0: R29 := 0
000001d5: R30 := 0
000001db: R5 := R0
000001e2: R1 := mem[R31, el]:u64
000001e8: R2 := R31 + 8
000001ee: R6 := R31
000001f3: R0 := 0x10000
000001fa: R0 := mem[R0 + 0xFF0, el]:u64
000001ff: R3 := 0
00000204: R4 := 0
00000209: R30 := 0x670
0000020c: call @__libc_start_main with return %0000020e

0000020e:
00000211: R30 := 0x674
00000214: call @abort with return %000005cb

000005cb:
000005cc: call @call_weak_fn with noreturn

000005ab: sub abort()


00000212:
00000468: R16 := 0x10000
0000046f: R17 := mem[R16 + 0xFC8, el]:u64
00000475: R16 := R16 + 0xFC8
0000047a: call R17 with noreturn

000005ac: sub call_weak_fn(call_weak_fn_result)
000005dd: call_weak_fn_result :: out u32 = low:32[R0]

00000216:
00000219: R0 := 0x10000
00000220: R0 := mem[R0 + 0xFE8, el]:u64
00000226: when R0 = 0 goto %00000224
000005cd: goto %00000384

00000224:
0000022c: call R30 with noreturn

00000384:
00000387: goto @__gmon_start__

00000385:
00000452: R16 := 0x10000
00000459: R17 := mem[R16 + 0xFC0, el]:u64
0000045f: R16 := R16 + 0xFC0
00000464: call R17 with noreturn

000005ae: sub deregister_tm_clones(deregister_tm_clones_result)
000005de: deregister_tm_clones_result :: out u32 = low:32[R0]

00000232:
00000235: R0 := 0x11000
0000023b: R0 := R0 + 0x10
00000240: R1 := 0x11000
00000246: R1 := R1 + 0x10
0000024c: #1 := ~R0
00000251: #2 := R1 + ~R0
00000257: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
0000025d: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
00000261: ZF := #2 + 1 = 0
00000265: NF := 63:63[#2 + 1]
0000026b: when ZF goto %00000269
000005ce: goto %00000366

00000366:
00000369: R1 := 0x10000
00000370: R1 := mem[R1 + 0xFD8, el]:u64
00000375: when R1 = 0 goto %00000269
000005cf: goto %00000379

00000269:
00000271: call R30 with noreturn

00000379:
0000037d: R16 := R1
00000382: call R16 with noreturn

000005b1: sub frame_dummy(frame_dummy_result)
000005df: frame_dummy_result :: out u32 = low:32[R0]

00000309:
0000030b: call @register_tm_clones with noreturn

000005b2: sub main(main_argc, main_argv, main_result)
000005e0: main_argc :: in u32 = low:32[R0]
000005e1: main_argv :: in out u64 = R1
000005e2: main_result :: out u32 = low:32[R0]

000001a8:
000001ab: call R30 with noreturn

000005b3: sub register_tm_clones(register_tm_clones_result)
000005e3: register_tm_clones_result :: out u32 = low:32[R0]

00000273:
00000276: R0 := 0x11000
0000027c: R0 := R0 + 0x10
00000281: R1 := 0x11000
00000287: R1 := R1 + 0x10
0000028e: R1 := R1 + ~R0 + 1
00000294: R2 := 0.63:63[R1]
0000029b: R1 := R2 + (R1 ~>> 3)
000002a1: R1 := extend:64[63:1[R1]]
000002a7: when R1 = 0 goto %000002a5
000005d0: goto %00000348

00000348:
0000034b: R2 := 0x10000
00000352: R2 := mem[R2 + 0xFF8, el]:u64
00000357: when R2 = 0 goto %000002a5
000005d1: goto %0000035b

000002a5:
000002ad: call R30 with noreturn

0000035b:
0000035f: R16 := R2
00000364: call R16 with noreturn
