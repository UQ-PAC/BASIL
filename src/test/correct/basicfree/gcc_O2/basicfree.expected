var {:extern} Gamma_mem: [bv64]bool;
var {:extern} mem: [bv64]bv8;
const {:extern} $_IO_stdin_used_addr: bv64;
axiom ($_IO_stdin_used_addr == 1896bv64);
function {:extern} memory_load8_le(memory: [bv64]bv8, index: bv64) returns (bv8) {
  memory[index]
}

procedure {:extern} rely();
  modifies Gamma_mem, mem;
  ensures (mem == old(mem));
  ensures (Gamma_mem == old(Gamma_mem));
  free ensures (memory_load8_le(mem, 1896bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 1897bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 1898bv64) == 2bv8);
  free ensures (memory_load8_le(mem, 1899bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69016bv64) == 80bv8);
  free ensures (memory_load8_le(mem, 69017bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69018bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69019bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69020bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69021bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69022bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69023bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69024bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69025bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69026bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69027bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69028bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69029bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69030bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69031bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69616bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69617bv64) == 6bv8);
  free ensures (memory_load8_le(mem, 69618bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69619bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69620bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69621bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69622bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69623bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69640bv64) == 8bv8);
  free ensures (memory_load8_le(mem, 69641bv64) == 16bv8);
  free ensures (memory_load8_le(mem, 69642bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 69643bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69644bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69645bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69646bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69647bv64) == 0bv8);

procedure {:extern} rely_transitive()
  modifies Gamma_mem, mem;
  ensures (mem == old(mem));
  ensures (Gamma_mem == old(Gamma_mem));
{
  call rely();
  call rely();
}

procedure {:extern} rely_reflexive();

procedure {:extern} guarantee_reflexive();
  modifies Gamma_mem, mem;

procedure main()
  free requires (memory_load8_le(mem, 69632bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69633bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69634bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69635bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69636bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69637bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69638bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69639bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69640bv64) == 8bv8);
  free requires (memory_load8_le(mem, 69641bv64) == 16bv8);
  free requires (memory_load8_le(mem, 69642bv64) == 1bv8);
  free requires (memory_load8_le(mem, 69643bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69644bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69645bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69646bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69647bv64) == 0bv8);
  free requires (memory_load8_le(mem, 1896bv64) == 1bv8);
  free requires (memory_load8_le(mem, 1897bv64) == 0bv8);
  free requires (memory_load8_le(mem, 1898bv64) == 2bv8);
  free requires (memory_load8_le(mem, 1899bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69016bv64) == 80bv8);
  free requires (memory_load8_le(mem, 69017bv64) == 7bv8);
  free requires (memory_load8_le(mem, 69018bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69019bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69020bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69021bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69022bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69023bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69024bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69025bv64) == 7bv8);
  free requires (memory_load8_le(mem, 69026bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69027bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69028bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69029bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69030bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69031bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69616bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69617bv64) == 6bv8);
  free requires (memory_load8_le(mem, 69618bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69619bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69620bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69621bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69622bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69623bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69640bv64) == 8bv8);
  free requires (memory_load8_le(mem, 69641bv64) == 16bv8);
  free requires (memory_load8_le(mem, 69642bv64) == 1bv8);
  free requires (memory_load8_le(mem, 69643bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69644bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69645bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69646bv64) == 0bv8);
  free requires (memory_load8_le(mem, 69647bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 1896bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 1897bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 1898bv64) == 2bv8);
  free ensures (memory_load8_le(mem, 1899bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69016bv64) == 80bv8);
  free ensures (memory_load8_le(mem, 69017bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69018bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69019bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69020bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69021bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69022bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69023bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69024bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69025bv64) == 7bv8);
  free ensures (memory_load8_le(mem, 69026bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69027bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69028bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69029bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69030bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69031bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69616bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69617bv64) == 6bv8);
  free ensures (memory_load8_le(mem, 69618bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69619bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69620bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69621bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69622bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69623bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69640bv64) == 8bv8);
  free ensures (memory_load8_le(mem, 69641bv64) == 16bv8);
  free ensures (memory_load8_le(mem, 69642bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 69643bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69644bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69645bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69646bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 69647bv64) == 0bv8);
{
  lmain:
    assume {:captureState "lmain"} true;
    goto main_return;
  terminate:
    assume {:captureState "terminate"} true;
    goto terminate;
  main_return:
    assume {:captureState "main_return"} true;
    return;
}
